graphql-language/src/parser.rs

```diff
@@ -0,0 +1,363 @@
+#![allow(dead_code, unused_variables)]
+use ast::*;
+use lexer::Punctuator::*;
+use lexer::Token::{
+  Eof,
+  Punctuator,
+  Name as TokenName,
+  IntValue as TokenInt,
+  FloatValue as TokenFloat,
+  StringValue as TokenString,
+};
+use lexer::{Lexer, Token};
+use source::Source;
+use std::iter::Peekable;
+
+pub type Res<'a, T> = Result<T, &'a str>;
+pub type DRes<'a, T> = Res<'a, (T, usize)>;
+
+pub struct Parser<'a> {
+  lexer: Peekable<Lexer<'a>>,
+  source: Source<'a>,
+}
+
+impl<'a> Parser<'a> {
+  pub fn new(src: Source<'a>) -> Self {
+    Parser {
+      lexer: Lexer::new(src.body).peekable(),
+      source: src,
+    }
+  }
+
+  fn loc(&self, lo: usize, hi: usize) -> Location<'a> {
+    Location {
+      start: lo,
+      end: hi,
+      source: self.source,
+    }
+  }
+}
+
+impl<'a> Iterator for Parser<'a> {
+  type Item = Res<'a, Definition<'a>>;
+  fn next(&mut self) -> Option<Res<'a, Definition<'a>>> {
+    if self.lexer.peek().is_some() {
+      Some(parse_definition(self))
+    } else {
+      None
+    }
+  }
+}
+
+macro_rules! on_do {
+ ($parser: ident, $($p: pat => $b: block)+) => {
+   loop {
+     match next!($parser) {
+       $(
+         $p => $b
+       )+
+     }
+   }
+ }
+}
+
+macro_rules! peek {
+  ($parser: ident, $($p: pat)|*) => ({
+    let mut is_match = false;
+    if let &Ok(ref c) = $parser.lexer.peek().unwrap() {
+      match c {
+        $(
+          &$p => {
+            is_match = true;
+          }
+        )*
+        _ => {}
+      }
+    }
+    is_match
+  });
+}
+
+macro_rules! next {
+  ($parser: ident) => {$parser.lexer.next().unwrap().unwrap()}
+}
+
+fn parse_definition<'a>(parser: &mut Parser<'a>) -> Res<'a, Definition<'a>> {
+  let token = next!(parser);
+
+  match token {
+    Punctuator(LeftBrace, lo, _) => parse_short_operation(parser, lo),
+    TokenName(name, lo, hi) => {
+      match name {
+        "query" => parse_operation_def(parser, OperationType::Query, name, lo, hi),
+        "mutation" => parse_operation_def(parser, OperationType::Mutation, name, lo, hi),
+        "fragment" => parse_fragment_def(parser, name, lo, hi),
+        _ => Err("Unkown Op"),
+      }
+    },
+    _ => Err("Unexpected Token"),
```

Comment:
This is why I don't like `&str` errors…best case scenario, we make the error a `String` instead, `impl` a location function on `Token`, and `impl Display for Token` so we could instead write:

```rust
token @ _ => Err(format!("Unexpected Token {} at {}.", token, token.loc().lo))
```

You could also look at the JS implementation which has [this](https://github.com/graphql/graphql-js/blob/master/src/language/parser.js#L225) nice method with the definition [here](https://github.com/graphql/graphql-js/blob/master/src/language/parser.js#L1094-L1105). If you move all your state into `Parser`, this would be super nice.

Better errors should be a blocker for this to get merged.